Plugin~/WebRTCPlugin/UnityRenderEvent.cpp

```diff
@@ -254,57 +260,76 @@ static void UNITY_INTERFACE_API OnRenderEvent(int eventID, void* data)
+    }
 
-    if (s_bufferPool->bufferCount() < kLimitBufferCount)
+    for (int i = 0; i < batchData->tracksCount; i++)
     {
-        std::unique_ptr<const ScopedProfiler> profiler;
-        if (s_ProfilerMarkerFactory)
-            profiler = s_ProfilerMarkerFactory->CreateScopedProfiler(*s_MarkerEncode);
+        VideoStreamTrackData* trackData = batchData->tracks[i];
+
+        if (trackData == nullptr || trackData->texture == nullptr || trackData->source == nullptr)
+            continue;
+
+        if (trackData->action == VideoStreamTrackAction::Encode)
+        {
+            RTC_DCHECK(trackData->texture);
+            RTC_DCHECK(trackData->source);
+            RTC_DCHECK_GT(trackData->width, 0);
+            RTC_DCHECK_GT(trackData->height, 0);
+
+            UnityVideoTrackSource* source = static_cast<UnityVideoTrackSource*>(trackData->source);
+            if (!s_context->ExistsRefPtr(source))
+            {
+                trackData->source = nullptr;
+                continue;
+            }
+
+            Timestamp timestamp = s_clock->CurrentTime();
+            IGraphicsDevice* device = Plugin::GraphicsDevice();
+            UnityGfxRenderer gfxRenderer = device->GetGfxRenderer();
+            void* ptr = GraphicsUtility::TextureHandleToNativeGraphicsPtr(trackData->texture, device, gfxRenderer);
+            unity::webrtc::Size size(trackData->width, trackData->height);
 
-        auto frame = s_bufferPool->CreateFrame(ptr, size, encodeData->format, timestamp);
-        source->OnFrameCaptured(std::move(frame));
+            if (s_bufferPool->bufferCount() < kLimitBufferCount)
+            {
+                std::unique_ptr<const ScopedProfiler> profiler;
+                if (s_ProfilerMarkerFactory)
+                    profiler = s_ProfilerMarkerFactory->CreateScopedProfiler(*s_MarkerEncode);
+
+                auto frame = s_bufferPool->CreateFrame(ptr, size, trackData->format, timestamp);
+                source->OnFrameCaptured(std::move(frame));
+            }
+        }
+#if 0
+        else if (trackData->action == VideoStreamTrackAction::Decode)
+        {
+            std::unique_ptr<const ScopedProfiler> profiler;
+            if (s_ProfilerMarkerFactory)
+                profiler = s_ProfilerMarkerFactory->CreateScopedProfiler(*s_MarkerDecodeCopy);
+            UnityVideoRenderer* renderer = static_cast<UnityVideoRenderer*>(trackData->source);
+            renderer->CopyBuffer(trackData->texture, trackData->width, trackData->height, trackData->format);
+        }
+#endif
     }
-    s_bufferPool->ReleaseStaleBuffers(timestamp, kStaleFrameLimit);
-}
 
-static void UNITY_INTERFACE_API OnReleaseBuffers(int eventID, void* data)
-{
-    if (eventID != s_releaseBuffersEventID)
-        return;
-    // Release all buffers.
-    if (s_bufferPool)
-        s_bufferPool->ReleaseStaleBuffers(Timestamp::PlusInfinity(), kStaleFrameLimit);
+    s_bufferPool->ReleaseStaleBuffers(timestamp, kStaleFrameLimit);
```

Comment:
I got this error.
```
C:\com.unity.webrtc\Plugin~\WebRTCPlugin\UnityRenderEvent.cpp(322,39): error : use of undeclared i
dentifier 'timestamp' [C:\com.unity.webrtc\Plugin~\out\build\x64-windows-clang\WebRTCPlugin\WebRTC
Plugin.vcxproj]
```

Reply:
Doh, last minute patch grafting error when porting over changes from another branch that doesn't use s_bufferPool. Fixed in https://github.com/Unity-Technologies/com.unity.webrtc/pull/915/commits/d29a0289305e247df2f69974dc99fab5aedf8aad